game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1216 @@
+
+    final List<StepUnits> units = new ArrayList<>(outcomes.values());
+
+    units.sort(
+        Comparator.comparingInt(StepUnits::countOfFriendlyDamagedOrDead)
+            .thenComparingInt(StepUnits::countOfFriendlyHitPoints)
+            .reversed());
+    return units;
+  }
+
+  private void getTargetGroupFightOutcomes(
+      final Map<StepUnits, StepUnits> outcomes,
+      final List<Triple<Collection<Unit>, Collection<Unit>, Boolean>> groupsAndTargets,
+      final int groupsAndTargetsIndex,
+      final StepUnits currentUnits,
+      final Type type,
+      final MustFightBattle.ReturnFire returnFire,
+      final boolean allowMultipleHitsPerUnit,
+      final double probability) {
+    final List<Unit> firingGroup =
+        new ArrayList<>(groupsAndTargets.get(groupsAndTargetsIndex).getFirst());
+    final Collection<Unit> aliveUnits = currentUnits.getAliveFriendly();
+    final Collection<Unit> validAliveOrInjuredTargets =
+        groupsAndTargets.get(groupsAndTargetsIndex).getSecond();
+    final List<Unit> validAliveTargets =
+        groupsAndTargets.get(groupsAndTargetsIndex).getSecond().stream()
+            .filter(aliveUnits::contains)
+            .collect(Collectors.toList());
+
+    final boolean isSuicideOnHit = groupsAndTargets.get(groupsAndTargetsIndex).getThird();
+
+    final boolean defending = type == Type.DEFENDER || type == Type.AA_DEFENDER;
+    final RollData diceTuple;
+
+    if (type == Type.AA_ATTACKER || type == Type.AA_DEFENDER) {
+      diceTuple =
+          getAaRollData(
+              currentUnits,
+              defending,
+              allowMultipleHitsPerUnit,
+              firingGroup,
+              validAliveOrInjuredTargets);
+    } else {
+      diceTuple = getRegularRollData(defending, firingGroup, validAliveOrInjuredTargets);
+    }
+    final List<Double> hitProbabilities = calculateHitProbabilities(diceTuple);
+
+    int totalHits = 0;
+    double totalProbability = 0.0;
+    for (final Unit targetUnit : currentUnits.addFriendlyMultiHitTargets(validAliveTargets)) {
+
+      if (totalHits >= diceTuple.totalRolls) {
+        // no more rolls available
+        break;
+      }
+
+      final double hitProbability = hitProbabilities.get(totalHits);
+      totalProbability += hitProbability;
+      if (groupsAndTargetsIndex > 0) {
+        getTargetGroupFightOutcomes(
+            outcomes,
+            groupsAndTargets,
+            groupsAndTargetsIndex - 1,
+            new StepUnits(currentUnits),
+            type,
+            returnFire,
+            allowMultipleHitsPerUnit,
+            hitProbability * probability);
+      } else {
+        outcomes.compute(
+            new StepUnits(currentUnits, hitProbability * probability),
+            (key, value) -> {
+              if (value == null) {
+                return key;
+              } else {
+                value.addProbability(hitProbability * probability);
+              }
+              return value;
+            });
```

Comment:
This is something I don't quite get.
Using this compute function the key would always be the same object as the value right?
So why bother using a map at all?

Reply:
The probability isn't part of the equality check.  So it allows summing up the probabilities if the other parts are equal.

Reply:
I assumed that already, thanks for clarifying that.
But my question still stands (even though I should've been more clear): Why use Map<StepUnit, StepUnit> over something like Set<StepUnit> or even Map<List<Unit>, Double>?

Reply:
I've changed it to Map<StepUnit, Double>